doc/developer/design/20230802_explain_running_dataflows.md

```diff
@@ -0,0 +1,255 @@
+```
+
+and implement similar behavior for
+
+```sql
+EXPLAIN [OPTIMIZED|PHYSICAL] PLAN FOR INDEX
+```
+
+# Explanation
+[explanation]: #explanation
+
+At the moment, `sequence_explain_plan` constructs its output using information derived from optimizations done at the time when `sequence_explain_plan` runs.
+
+At the high level, the API that constructs an `EXPLAIN` output for a given plan looks like that:
+
+```rust
+let context = ExplainContext {
+    config,
+    humanizer,
+    used_indexes,
+    finishing,
+    duration,
+};
+
+Explainable::new(&mut plan).explain(format, context)
+```
+
+We need to
+(1) analyze which values in the above code snippet represent information that should reflect the state of the world when the dataflow was originally created, and
+(2) suggest mechanisms to make these values available for subsequent `EXPLAIN <catalog item>` calls.
+
+In general, we can persist the required data
+(a) in-memory, as extensions of the `CatalogItem` objects that are maintained as entries in the `CatalogState`, or
+(b) durably, as extensions of the `SerializedCatalogItem` structures that represent the part of the catalog state that survives restarts.
+
+Since at the moment every `CatalogItem` is fully re-optimized upon an `environmentd` restart in the catalog re-hydration phase, the current proposal focuses on option (a).
+However, note that if we introduce plan pinning or similar features in the future, we might need to revisit this decision.
+
+The comments in the following snippet summarize the high-level implementation plan:
+
+```rust
+let context = ExplainContext {
+    config: &config,     // Derived from the enclosing `EXPLAIN` statement plan.
+    humanizer: &catalog, // Passing the current catalog state.
+    used_indexes,        // Passing a `CatalogItem` field.
+    finishing,           // Always use `None` here.
+    duration,            // Always use `Duration::default()` here.
+};
+
+// Constructed using values from a `CatalogItem` field.
+// Use `format` from the enclosing `EXPLAIN` plan.
+Explainable::new(&mut plan).explain(format, context)
+```
+
+Basically
+
+- The `config` is something only depends on the current catalog state.
+- The `humanizer` value can be the current catalog state because item IDs are stable and cannot be recycled.
+- The `finishing` and `duration` values an be constants.
+
+so the only time-dependent information that we need to record is the `plan` and the `used_indexes`.
+The next section expands how we plan to do that with more concrete implementation details.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+## Extend `CatalogItem` variants
+
+Extend and make an extra field in `Index` and `MaterializedView` with the following fields:
+
+```rust
+// idential extensions to `index`
+struct MaterializedView {
+    ...
+    optimized_plan: Box<mz_expr::MirRelationExpr>,
+    physical_plan: Box<mz_compute_client::Plan>,
+    used_indexes: UsedIndexes,
```

Comment:
What does the `UsedIndexes` struct look like? Also is that information derivable from the optimized or physical plan?

Reply:
What happens if one of the indexes in `UsedIndexes` is deleted? Then `UsedIndexes` will contain a dangling reference. Or is that ok because we never actually try and look the index up?

Reply:
I need to dig the place of the code where this is handled, but my understanding is that:

1. We will delete the catalog entry.
2. The dataflow that maintains the index will continue do exist as long as it has dependencies.
3. When we restart `environmentd`, we will re-optimize and re-isntall the materialized view, but at that point the deleted `GlobalId` will not land in the updated `used_indexes`.

Reply:
Btw. this is the issue about the problem that currently you are allowed to DROP an index that is in use, resulting in the weird behavior that Alex explained above: https://github.com/MaterializeInc/materialize/issues/20334

Reply:
At the moment its this (instead of printing the name I just print a dangling `GlobalId`):

```text
materialize=> explain materialized view mv;
                            Optimized Plan                            
----------------------------------------------------------------------
 t2:                                                                 +
   Project (#0..=#2)                                                 +
     ReadExistingIndex materialize.public.accounts lookup_value=(100)+
                                                                     +
 Used Indexes:                                                       +
   - materialize.public.accounts_balance_idx (lookup)                +
 
(1 row)

materialize=> drop index accounts_balance_idx;
DROP INDEX
materialize=> explain materialized view mv;
                            Optimized Plan                            
----------------------------------------------------------------------
 t2:                                                                 +
   Project (#0..=#2)                                                 +
     ReadExistingIndex materialize.public.accounts lookup_value=(100)+
                                                                     +
 Used Indexes:                                                       +
   - u9 (lookup)                                                     +
 
(1 row)
```

I can change this to print

```text
 t2:                                                                 +
   Project (#0..=#2)                                                 +
     ReadExistingIndex materialize.public.accounts lookup_value=(100)+
                                                                     +
 Used Indexes:                                                       +
   - materialize.public.accounts_balance_idx [deleted] (lookup)      +
```